### PR TITLE
Release v0.0.118: string utilities + character classification (#470, #471)

### LINKED ISSUE #470: Add string utility built-ins (reverse, pad, chars, trim variants)
## Summary

Vera's string built-ins cover the most common operations but are missing several utilities that come up frequently in text processing and formatting. Two distinct gaps:

1. **Character-level ops** (`string_reverse`, `string_pad_*`, `string_chars`, `string_trim_*`) — close the per-character recursion pattern.
2. **Structural splits** (`string_lines`, `string_words`) — `string_split` only takes a single delimiter character, so splitting on line endings (handling `\r\n`) or whitespace runs requires regex today.

### Transformation

- `string_reverse(@String) -> @String` — reverse a string (Unicode-aware, by grapheme cluster)
- `string_pad_start(@String, @Nat, @String) -> @String` — pad to target length with fill string (left-pad)
- `string_pad_end(@String, @Nat, @String) -> @String` — pad to target length with fill string (right-pad)

### Decomposition

- `string_chars(@String) -> @Array<String>` — split into individual characters (grapheme clusters) as an array of single-character strings. **Bridge primitive**: once this exists, string operations become array operations (e.g. `string_reverse = string_from_chars(array_reverse(string_chars(s)))`, character predicates become `array_any(string_chars(s), is_digit)`).

### Structural splits

- `string_lines(@String) -> @Array<String>` — split on line terminators (`\n`, `\r\n`, `\r`). Mirrors Python's `str.splitlines()`, Rust's `str::lines()`, Java's `String.lines()`. Current workaround requires `string_split("\n")` and manual `\r` stripping — LLMs get this wrong on Windows-style line endings.
- `string_words(@String) -> @Array<String>` — split on whitespace runs (spaces, tabs, newlines), discarding empty segments. Mirrors Python's `str.split()` with no argument. Currently impossible with `string_split` (single-delimiter only); requires regex.

### Trimming variants

- `string_trim_start(@String) -> @String` — remove leading whitespace only
- `string_trim_end(@String) -> @String` — remove trailing whitespace only

Note: `string_strip` already exists for trimming both sides.

## Implementation

- **environment.py**: Register as pure functions
- **codegen/api.py**: Host imports delegating to Python string methods (`.ljust()`, `.rjust()`, `.lstrip()`, `.rstrip()`, `list()` for chars, `.splitlines()` for lines, `.split()` with no arg for words)
- **Browser runtime**: `padStart()`, `padEnd()`, `trimStart()`, `trimEnd()`, spread operator for chars, regex-based split for lines/words
- **Verification**:
  - `string_reverse` preserves length
  - `string_pad_start` / `string_pad_end`: result length >= input length
  - `string_trim_*`: result length <= input length
  - `string_chars`: result length = `string_length` of input
  - `string_lines` / `string_words`: concatenating results (with separator) recovers a prefix of the input

## Priority

Medium. Less urgent than array utilities, sleep, and random. Within this issue, priority order:

1. **`string_chars`** — bridge primitive; unlocks array-combinator approach to string processing
2. **`string_lines`** / **`string_words`** — real capability gaps; `string_split` can't currently do these correctly
3. **`string_pad_start`** / **`string_pad_end`** — common formatting need
4. **`string_reverse`**, **`string_trim_start`**, **`string_trim_end`** — convenience, once `string_chars` exists several of these become one-liners

### LINKED ISSUE #471: Add character classification built-ins (is_digit, is_alpha, etc.)
## Summary

Vera has `string_char_code` and `string_from_char_code` for working with individual characters, but no character classification or case-conversion functions for single characters. These are fundamental for parsing, validation, and text processing — and hand-rolling the equivalents via ASCII range comparisons is a common source of off-by-one errors for LLMs.

## Proposed API

All functions take a `String` (first character is examined) and return `Bool` or `String`:

### Classification (boolean)

- `is_digit(@String) -> @Bool` — ASCII 0–9
- `is_alpha(@String) -> @Bool` — ASCII a–z, A–Z
- `is_alphanumeric(@String) -> @Bool` — ASCII letter or digit
- `is_whitespace(@String) -> @Bool` — space, tab, newline, carriage return
- `is_upper(@String) -> @Bool` — ASCII A–Z
- `is_lower(@String) -> @Bool` — ASCII a–z

All classification functions return `false` for empty strings.

### Case conversion (single-character)

- `char_to_upper(@String) -> @String` — uppercase the first character; return other characters unchanged.
- `char_to_lower(@String) -> @String` — lowercase the first character; return other characters unchanged.

Both return the empty string unchanged. Operate on the first character of the input; callers can slice if they want whole-string conversion (`string_upper` / `string_lower` already exist for that).

### Rationale for adding single-character case conversion

Whole-string `string_upper` / `string_lower` already exist. But the common "capitalize the first letter of a word" pattern currently requires:

```vera
string_concat(
  string_upper(string_slice(s, 0, 1)),
  string_slice(s, 1, string_length(s))
)
```

That's verbose and every `string_slice` offset is an error surface. `char_to_upper` / `char_to_lower` combine naturally with `string_chars` (once [#470](https://github.com/aallan/vera/issues/470) lands) for building case-transformation pipelines.

## Implementation

- **environment.py**: Register as pure functions — `(STRING,) -> BOOL` for classifiers, `(STRING,) -> STRING` for case conversion
- **codegen/api.py**: Host imports using Python's `str.isdigit()`, `str.isalpha()`, etc. on the first character; `s[0].upper() + s[1:]` (and `.lower()`) for the case converters
- **Browser runtime**: Regex or char-code range checks for classification; `s[0].toUpperCase() + s.slice(1)` for case conversion
- **Verification**: Boolean predicates — standard Z3 encoding. Can relate to `string_char_code` axioms (e.g. `is_digit(s) <=> 48 <= string_char_code(s, 0) <= 57`). Case conversion: result length equals input length.

## Design note

These operate on the first character of a string rather than taking a char type, because Vera has no `Char` type — characters are represented as single-character strings (same as Elm, PureScript). This is consistent with `string_char_code(@String, @Nat) -> @Int` which also operates on a string position.

## Priority

Lower priority than array utilities, sleep, and random. Within this issue: classification functions and case converters ship together (same implementation pattern, same design note). Useful for hand-written parsers, input validation, and the first-letter-capitalize pattern.

### PR DESCRIPTION
**Pull request by aallan**
## Summary

Sixteen new built-in functions for string utilities (#470) and character classification (#471), shipped together as a single PR. All inline WAT — no host imports, bit-identical Python (wasmtime) and browser (Node.js) output.

**String utilities (8, #470):**

- `string_chars` / `string_lines` / `string_words` — bridge from `String` to `Array<String>`. Lines follows Python `splitlines()` (trailing terminator does not add empty segment); words follows Python `split()` with no args (whitespace runs collapse, empty segments discarded).
- `string_reverse` / `string_trim_start` / `string_trim_end`
- `string_pad_start` / `string_pad_end` — JavaScript `padStart` / `padEnd` semantics: fill cycles left-to-right, truncated to padding length; empty fill is a no-op; target shorter than input returns input unchanged.

**Character classification (8, #471):**

- `is_digit` / `is_alpha` / `is_alphanumeric` / `is_whitespace` / `is_upper` / `is_lower` — first-byte ASCII predicates; all return `false` for the empty string.
- `char_to_upper` / `char_to_lower` — first byte only.

ASCII-only by design. Phase 2 (Unicode-aware variants — `string_codepoints`, `string_graphemes`, whole-string case conversion, etc.) tracked separately in #509.

## Implementation

- `vera/wasm/calls_strings.py` grew from 2,588 → 3,979 lines with sixteen new translators. Shared helpers factor out the predicate scaffolding (`_translate_classifier`), case conversion (`_translate_char_case`), trim variants (`_translate_trim`), pad variants (`_translate_pad`), and structural splits (`_translate_structural_split` covers both `string_lines` and `string_words` via a two-pass count-then-emit shape mirroring the existing `string_split`).
- ASCII range tricks reused throughout: `(byte - 48) < 10` for digit, `((byte | 0x20) - 97) < 26` for case-folded alpha. Branchless and idiomatic.
- `vera/wasm/inference.py` extended for the new return-type shapes (`i32_pair` for string/array returns, `i32` for bool returns).
- `vera/wasm/calls.py` dispatch table extended with sixteen new branches; `_infer_concat_elem_type` now returns `"String"` for the three split operations.
- `vera/codegen/modules.py` known-names allowlist extended.
- `vera/environment.py` registers all sixteen `FunctionInfo` entries.

## Tests

- **33 unit tests** across `TestCharClassification` and `TestStringUtilities` in `tests/test_codegen.py`.
- **9 browser parity tests** across `TestBrowserStringUtilities` and `TestBrowserCharClassification` in `tests/test_browser.py`.
- **New conformance program** `tests/conformance/ch09_string_char_utilities.vera` (level: `run`) with 14 test functions covering classifier, transform, pad, split, and empty-string edge cases.
- **Tier-count test** updated: 184/23/207 → 213/26/239 reflects the new `examples/string_utilities.vera` contributing 29 T1 + 3 T3 + 32 contracts.

## Example

`examples/string_utilities.vera` exercises all sixteen built-ins in a coherent log-line-processor scenario — line counting, word splitting, byte iteration via `string_chars`, classifier-based token validation, title-casing, trim, pad, palindrome check via `string_reverse`, plus comprehensive empty-string round-trip coverage. Verified clean (29 Tier 1, 3 Tier 3).

## Documentation

Comprehensive docs sweep across both human- and agent-facing surfaces:

- **spec/09-standard-library.md** — new §9.6.22 with full signatures, semantics, ASCII range tables, and verification notes.
- **SKILL.md** — new sub-section under string ops; naming-convention section mentions the sixteen new names; bumped conformance count to 79.
- **AGENTS.md / CLAUDE.md / README.md / vera/README.md / FAQ.md / TESTING.md / EXAMPLES.md** — count bumps (78 → 79 conformance, 31 → 32 examples, 137 → 153 built-ins, 3,437 → 3,489 tests), file-size table refresh.
- **CHANGELOG.md** — v0.0.118 entry with separate Added / Compiler / Tests sections.
- **HISTORY.md** — Stage 11 row added.
- **ROADMAP.md** — #470 / #471 removed from short-term queue (now shipped); #509 Unicode phase-2 added to Phase 4c.
- **examples/README.md** — `string_utilities.vera` row added.
- **docs/** site assets regenerated via `scripts/build_site.py`.

## Release prep

- Version bumped 0.0.117 → 0.0.118 across `vera/__init__.py`, `pyproject.toml`, `docs/index.html`, `README.md`, `uv.lock`.
- Tag `v0.0.118` placed on the feature branch; will be moved to main HEAD after merge so the CHANGELOG-extraction release flow can run.

## Test plan

- [x] 33 new unit tests pass (`pytest tests/test_codegen.py::TestCharClassification tests/test_codegen.py::TestStringUtilities`)
- [x] 9 new browser parity tests pass (`pytest tests/test_browser.py::TestBrowserStringUtilities tests/test_browser.py::TestBrowserCharClassification`)
- [x] All 79 conformance programs pass (`python scripts/check_conformance.py`)
- [x] All 32 examples pass check + verify (`python scripts/check_examples.py`)
- [x] Full pytest suite green (3,475 passed, 14 skipped)
- [x] mypy clean
- [x] All 14 doc validation scripts green (conformance, examples, README, spec, SKILL, EXAMPLES, FAQ, HTML, site assets, version sync, doc counts, licenses, limitations, examples README)

Closes #470
Closes #471

## Summary by CodeRabbit

* **New Features**
  * Added 16 built-in string and character utilities (chars/lines/words, reverse, trim start/end, pad start/end, ASCII first-byte classifiers and case conversions).

* **Tests**
  * Expanded unit and browser-parity tests and added a new conformance chapter covering the string/char utilities; verification totals updated.

* **Documentation**
  * Updated changelog, READMEs, roadmap, specs and other docs to v0.0.118 and revised test/conformance/example counts.

* **Chores**
  * Bumped package version to v0.0.118.

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -252,6 +252,7 @@ Stage 11 shifts focus from evaluation infrastructure to the standard library and
...
...single PR, all inline WAT (no host imports, bit-identical between `wasmtime` and Node.js). String utilities: `string_chars` (canonical bridge from `String` to `Array<String>`; each 1-byte element shares a single backing data buffer), `string_lines` (Python `splitlines()` — `\n` / `\r\n` / `\r` terminators; trailing terminator does not add empty segment), `string_words` (Python `split()` — whitespace runs collapse, empty segments discarded), `string_reverse`, `string_trim_start`, `string_trim_end`, `string_pad_start`, `string_pad_end` (JavaScript `padStart`/`padEnd` — fill cycles left-to-right and is truncated to exactly the padding length; empty fill is a no-op). Character classification: `is_digit`, `is_alpha`, `is_alphanumeric`, `is_whitespace`, `is_upper`, `is_lower` (first-byte ASCII predicates; `false` for empty string). First-character case conversion: `char_to_upper`, `char_to_lower` (only the first byte is transformed; remaining bytes pass through — useful for title-casing). ASCII-only; Unicode-aware variants tracked separately. Eliminates the hand-rolled `byte == 48 || byte == 49 || ...` patterns and gives agents the string primitives that every non-trivial parser needs. |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Both fixed in commit 4100790 — the literal `||` is now `or`, and the footer count is bumped to **118** (the live total per the link refs at the bottom of CHANGELOG.md, not 114 as suggested — HISTORY.md was actually 5 stale, not 1).

File: scripts/check_skill_examples.py [hunk 2]
```diff
@@ -79,77 +79,80 @@
...
     # Float64 predicates — bare function calls (shifted by De Bruijn section additions)
-    933: ("FRAGMENT", "Float64 predicate and constant examples, bare calls"),
+    967: ("FRAGMENT", "Float64 predicate and constant examples, bare calls"),
 
     # Contracts incremental workflow — placeholder scaffolding, not a full function
-    1001: ("FRAGMENT", "Contracts scaffolding template: requires(true) ensures(true)"),
+    1035: ("FRAGMENT", "Contracts scaffolding template: requires(true) ensures(true)"),
+
+    # String utilities and character classification — bare function-call signatures
+    791: ("FRAGMENT", "String utility / classifier signatures, bare function calls"),
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_skill_examples.py:155]: @coderabbitai Verified, no fix needed. Line 791 IS the actual line of the opening ```vera fence in SKILL.md after my insertion, and `python scripts/check_skill_examples.py` exits 0 locally (50 parsed OK, 57 allowlisted, 0 failed). The fragment is being skipped correctly.

File: vera/README.md [hunk 3]
```diff
@@ -104,7 +104,7 @@ execute(compile_result, ...)    # → run WASM via wasmtime
 | ` ├ calls_markup.py` | 315 | | JSON, HTML, Markdown, Regex, async/await (host-import wrappers) | |
 | ` ├ calls_math.py` | 457 | | `abs`, `min`, `max`, `floor`, `ceil`, `round`, `sqrt`, `pow`, Float64 predicates, numeric conversions | |
 | ` ├ calls_parsing.py` | 970 | | `parse_nat` / `parse_int` / `parse_bool` / `parse_float64` state machines | |
-| ` ├ calls_strings.py` | 2,588 | | All string ops (length, concat, slice, search, transform, split, join) + to-string conversions | |
+| ` ├ calls_strings.py` | 3,979 | | All string ops (length, concat, slice, search, transform, split, join, chars/lines/words, reverse, trim_start/end, pad_start/end, char_to_upper/lower, classifiers) + to-string conversions | |
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/README.md]: @coderabbitai Fixed in commit 4100790, bumped to **16,608** rather than 16,098. The 16,608 figure includes the further line growth from the per-slice-allocation refactor in `_translate_structural_split` (a separate CR finding addressed in the same commit) — `calls_strings.py` is now 3,977 lines, and `wc -l vera/wasm/*.py` confirms the package total.

**Comment by aallan** [comment]
@coderabbitai Round 2 addressed in commit 98f3a74. Verified each finding against the current code; fixed seven, dismissed two as already-correct, treated one as a duplicate of the previous round.

**Fixed:**
- Stale "shared backing data buffer" claim corrected in CHANGELOG.md, HISTORY.md, spec/09-standard-library.md §9.6.22, the calls_strings.py block comment, and the TestStringUtilities docstring (the per-slice-allocation refactor in 4100790 made all five wordings stale).
- Added `test_char_to_lower_empty` mirror, `test_string_pad_end_no_change_when_longer`, `test_string_pad_end_empty_fill`, `test_string_trim_vt_ff_full_set`, `test_string_words_vt_ff_separators`, `test_string_words_only_vt_ff` to test_codegen.py. VT/FF spelled with `\u{0B}` / `\u{0C}` since Vera's lexer doesn't accept `\v` / `\f` simple escapes.
- Extended browser `test_string_pad` (empty-fill no-op for both sides), `test_string_trim` (VT/FF mixed in), and `test_string_words` (VT/FF separators + all-VT/FF input).
- **Real bug:** `_translate_strip` was open-coding a narrower whitespace predicate {tab, LF, CR, space} that diverged from `_translate_trim` after the #470 series added VT and FF. Refactored `_translate_strip` to delegate to `_translate_trim(trim_start=True, trim_end=True)` — eliminates ~160 lines of duplicate code and locks the predicate in sync.
- vera/README.md aggregate refreshed (was duplicate of previous-round comment but the strip refactor changed sizes again — calls_strings.py now 3,890 lines, wasm/ total 16,521).

**Dismissed:**
- spec/09 fence convention: existing chapter-9 convention is `vera`-tagged + FRAGMENT allowlist (matches string_index_of/replace/join/repeat at lines 1108/1190/1221/1250). My five new blocks already follow it.

3,486 tests pass, mypy clean, all 14 doc validation scripts green.

File: spec/09-standard-library.md [hunk 4]
```diff
@@ -1671,6 +1671,143 @@ The callback-based combinators — `array_mapi`, `array_find`, `array_any`, `arr
...
+```vera
+public fn string_chars(@String -> @Array<String>)
+  requires(true) ensures(true) effects(pure)
+
+public fn string_lines(@String -> @Array<String>)
+  requires(true) ensures(true) effects(pure)
+
+public fn string_words(@String -> @Array<String>)
+  requires(true) ensures(true) effects(pure)
+```
```
[the remark this answers is not in the document]
Comment (aallan) [reply spec/09-standard-library.md:1691]: @coderabbitai Thanks. The strip → trim delegation also flagged a category lesson worth carrying forward: when two functions share a predicate (here: "is ASCII whitespace"), routing both through one helper from day one would have prevented this entire round of work. I'll keep an eye out during the upcoming Stage 11 builtins (#366, #507) for similar predicate-duplication risks — anywhere two builtins inspect the same byte set, one should call the other rather than re-encode the rule.

[the remark this answers is not in the document]
Comment (aallan) [reply spec/09-standard-library.md:1691]: @coderabbitai That heuristic is sharp — "overlapping i32.const byte literals = candidate for delegation" is exactly the kind of mechanical check I can apply at PR review time without needing to remember the broader category rule. Filing it in my head as the **byte-literal-collision rule** alongside the existing GC-shadow-push rule and the inline-WAT-no-host-imports rule.

File: TESTING.md [hunk 5]
```diff
@@ -6,10 +6,10 @@ This is the single source of truth for Vera's testing infrastructure, coverage d
 
 | Metric | Value |
 |--------|-------|
-| **Tests** | 3,437 across 28 files (~34,000 lines of test code; 3,423 passed, 14 skipped) |
+| **Tests** | 3,500 across 28 files (~34,000 lines of test code; 3,486 passed, 14 skipped) |
 | **Compiler code coverage** | 96% of 15,149 statements (CI minimum: 80%) |
-| **Conformance programs** | 78 programs across 9 spec chapters, validating every language feature |
-| **Example programs** | 31, all validated through `vera check` + `vera verify` |
+| **Conformance programs** | 79 programs across 9 spec chapters, validating every language feature |
+| **Example programs** | 32, all validated through `vera check` + `vera verify` |
```
[the remark this answers is not in the document]
Comment (aallan) [reply TESTING.md:12]: @coderabbitai Fixed in commit fb4d0b6 — line 274 now reads "all 32 examples".  `python scripts/check_doc_counts.py` exits 0.

File: tests/test_browser.py [hunk 6]
```diff
@@ -976,6 +976,256 @@ def test_array_sort_by_stability(self, tmp_path: Path) -> None:
...
+  IO.print(",");
+  IO.print(string_join(string_lines("a\\r\\nb\\rc"), "|"));
+  IO.print(",");
+  IO.print(int_to_string(nat_to_int(array_length(string_lines("a\\n")))))
+}
+'''
+        wasm_path, _ = _compile_vera(source, tmp_path)
+        node = _run_node(wasm_path)
+        # Trailing newline does NOT create empty final segment.
+        assert node["stdout"] == "a|b|c,a|b|c,1"
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_browser.py]: @coderabbitai Fixed in commit fb4d0b6 — extended the source program with `string_lines("")` and updated the expected stdout to include the trailing `,0`.  Now exercises the alloc(0) / zero-segment branch under the browser runtime.

File: vera/README.md [hunk 7]
```diff
@@ -91,7 +91,7 @@ execute(compile_result, ...)    # → run WASM via wasmtime
 | `resolver.py` | 213 | Resolve | Module path resolution, parse cache | `ModuleResolver` |
 | `smt.py` | 1,026 | Verify | Z3 translation layer | `SmtContext`, `SlotEnv` |
 | `verifier.py` | 1,005 | Verify | Contract verification | `verify()` |
-| `wasm/` | 12,998 | Compile | WASM translation layer (package) | `WasmContext`, `WasmSlotEnv`, `StringPool` |
+| `wasm/` | 16,521 | Compile | WASM translation layer (package) | `WasmContext`, `WasmSlotEnv`, `StringPool` |
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/README.md:94]: @coderabbitai Fixed in commit fb4d0b6 — line 486 now reads `wasm/ (16,521 lines across 18 modules)`, matching the table at line 94.  Will keep an eye out for this duplicated-figure pattern; ideally the README would derive both from a single source.

File: vera/wasm/calls_strings.py [hunk 8]
```diff
@@ -2586,3 +2451,1440 @@ def _translate_join(
...
+
+        ins.append(f"local.get {end}")
+        ins.append(f"local.get {start}")
+        ins.append("i32.sub")
+        ins.append(f"local.set {new_len}")
+
+        ins.append(f"local.get {new_len}")
+        ins.append("call $alloc")
+        ins.append(f"local.set {dst}")
+        ins.extend(gc_shadow_push(dst))
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_strings.py:2962]: @coderabbitai Fixed in commit fb4d0b6 — `_translate_trim` now `gc_shadow_push(ptr)`s before the destination `$alloc`, with a comment explaining the rationale.  This was a textbook case of the **byte-literal-collision rule and the GC-root rule being independent**: my round-2 strip→trim refactor consolidated the duplicate predicate (rule 1) but didn't audit the surviving copy for missing roots (rule 2).  Saved that corollary to my review heuristics — DRY refactors don't audit for orthogonal bugs.  Verified `_translate_string_chars` already roots its `ptr_s`; the missing root was unique to `_translate_trim`.

File: tests/test_browser.py [hunk 9]
```diff
@@ -976,6 +976,266 @@ def test_array_sort_by_stability(self, tmp_path: Path) -> None:
...
+  -- VT/FF mixed in with regular whitespace.
+  IO.print(string_trim_start(" \\u{0B}\\u{0C}hi  "));
+  IO.print("|");
+  IO.print(string_trim_end("  hi\\u{0B}\\u{0C} "));
+  IO.print("|")
+}
+'''
+        wasm_path, _ = _compile_vera(source, tmp_path)
+        node = _run_node(wasm_path)
+        assert node["stdout"] == "hi  |  hi|hi  |  hi|"
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_browser.py:1029]: @coderabbitai Fixed in commit 407772b — added `test_string_strip_vt_ff` immediately after `test_string_trim`.  The test calls `string_strip("\u{0B}\u{0C}hi \u{0B}")` and asserts `"hi"`, pinning the strip→trim delegation contract under the browser runtime so a future refactor that re-opens the old narrow predicate for strip would break the assertion.

File: vera/wasm/calls_strings.py [hunk 10]
```diff
@@ -2586,3 +2451,1449 @@ def _translate_join(
...
+            # ASCII whitespace per Python's str.isspace():
+            # {tab(9), LF(10), VT(11), FF(12), CR(13), space(32)}.
+            # 9..=13 collapses to (byte - 9) < 5 for branchless emit.
+            return [
+                f"    local.set {byte}",
+                f"    local.get {byte}", "    i32.const 32", "    i32.eq",
+                f"    local.get {byte}", "    i32.const 9",  "    i32.sub",
+                "    i32.const 5", "    i32.lt_u",
+                "    i32.or",
+            ]
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_strings.py:2906]: @coderabbitai Fixed in commit 407772b — extracted `emit_is_ascii_whitespace(byte_local, indent="")` into `vera/wasm/helpers.py` alongside `gc_shadow_push`, replaced all four open-coded sites (`_translate_is_whitespace`, the `_is_ws_inline` closure inside `_translate_trim`, and both passes of `_translate_structural_split`) with calls to it.  The helper docstring captures the historical context (PR #510 round 2 caught `_translate_strip`'s divergent narrow predicate; round 4 hoisted the helper) and explicitly warns against re-encoding the byte set.  Direct application of the byte-literal-collision review rule — the cost of one helper is essentially zero, and silent divergence is now structurally impossible for this predicate.
